### PR TITLE
Allow using DEVICE_DT_DEFINE from C++ code

### PR DESCRIPTION
**Pull request by ewalt1**
Signed-off-by: Nick Ewalt <nicholasewalt@google.com>

### CODE REVIEW

**Review by dcpleung — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]

**Comment by cfriedt** [comment]
@ewalt1 - can you please file an issue (bug) and link it with this PR? Please don't use the "Fixes" mechanism though; we can merge the PR, but leave an issue open to revisit whenever this bug has been fixed upstream or in the Zephyr SDK.

In the issue, can you also be somewhat specific about the GCC version used, etc?

cc @stephanosio, @galak

File: include/device.h [hunk 1]
```diff
@@ -791,10 +791,18 @@ static inline bool device_is_ready(const struct device *dev)
  * `gen_handles.py` must be updated.
  */
 BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+#ifdef __cplusplus
+/* Workaround for GCC bug 83271.
```
Comment (stephanosio) [inline include/device.h:795]: Please add FIXME or TODO here so that it can be easily grepped in the future.

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by stephanosio** [comment]
> In the issue, can you also be somewhat specific about the GCC version used, etc?

I can see that this issue is still present in the GCC version used by the Zephyr SDK (GCC 10.3.0).

> whenever this bug has been fixed upstream

The problem is that this is not really a "bug," but instead an undefined behaviour (C++ standard does not dictate how this case should be handled).

At most, this is GCC making a different interpretation of the C++ standard to that of the Clang; so, I highly doubt this will ever be "fixed" upstream.

**Review by tbursztyka — DISMISSED** [review]
@ewalt1  can you work on @stephanosio's comment?

**Review by stephanosio — DISMISSED** [review]

**Review by mbolivar-nordic — DISMISSED** [review]
Looks like this slipped through the cracks. @ewalt1 could you please rebase? @stephanosio please re-approve when the rebase comes in

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> @ewalt1 could you please rebase?

Ugh, I'm sorry to do this to you again, but it looks like you've got another rebase conflict

**Comment by ewalt1** [comment]
It shows that there are no merge conflicts in my repo, I don't know github well enough to understand what the issue is. Feel free to open a different PR with the same fix.

**Pull request closed without merging**